### ISSUE TITLE
Rats example does not work, problem with dimension of x

### ISSUE THREAD

**Issue by josephwb** [stan-dev/rstan#204]
Running the example from the [documentation](https://github.com/stan-dev/rstan/wiki/RStan-Getting-Started#example-2-rats) gets:

``` r
rats_fit <- stan(file='rats.stan', verbose=FALSE);
COMPILING THE C++ CODE FOR MODEL 'rats' NOW.
Error : mismatch in dimension declared and found in context; processing stage=data initialization; variable name=x; position=0; dims declared=(1); dims found=(5)
failed to create the sampler; sampling not done
```

Possibly related to https://github.com/stan-dev/example-models/issues/25?

**Comment by jgabry** [comment]
This doesn't seem related to that other issue since that one is just about changing the prior distributions. Here it looks like it's not finding the data properly. Does it work if you explicitly tell it what the data is using the code below?

`rats_fit <- stan(file='rats.stan', data = list(N=N, T=T, y=y, x=x, xbar=xbar))`

**Comment by josephwb** [comment]
Doh. Yes, that is obviously the problem: `data` was not specified. This looks like a simple fix to the documentation.

**Comment by jgabry** [comment]
Actually the way you had it originally should work because if you omit the data argument it should look for the objects in your global environment. 

I just realized why this error is happening. The code on the wiki page is this:

```
y <- read.table('rats.txt', header = TRUE)
x <- c(8, 15, 22, 29, 36)
xbar <- mean(x)
N <- nrow(y)
T <- ncol(y)
rats_fit <- stan(file = 'rats.stan', verbose = FALSE)
```

Because there's a variable `T` in the `rats.stan` file we've defined a variable `T` in the R global environment. In this case `T <- ncol(y)`, which is `T = 5`. If the `data` argument is not a named list (i.e., if it is omitted or if it is a character vector of names) then when `stan` goes to search for the variables it goes and looks for `T` and decides that `T` is `TRUE` rather than the object `T` we created which has a value of 5.

**Comment by jgabry** [comment]
@bgoodri @maverickg I'm not sure how the `stan` function searches for the variables, but is there a way to have it find the right `T` without it having to be a named list? I'm surprised more people haven't had this problem given how often `T` tends to be used in .stan programs.

**Comment by bgoodri** [comment]
It looks as if its scoping is wrong.

**Comment by jgabry** [comment]
Yeah, this line in the `parse_data` function

```
mget(objects, envir = e, inherits = TRUE, ifnotfound = vector("list", length(objects)))
```

returns a list in which `T` is `TRUE` instead of the right value.

**Comment by bgoodri** [comment]
Fixed by 776e4f4, although I don't really understand why it was picking up the wrong things in the first place.

**Comment by Shwoobly** [comment]
Hi, I got a parsing error when running the rats_fit code.
The 8 schools example worked just fine. Using a Macbook M2.

PARSER FAILED TO PARSE INPUT COMPLETELY
STOPPED AT LINE 1: 
/ http://www.mrc-bsu.cam.ac.uk/bugs/winbugs/Vol1.pdf
// Page 3: Rats
data {
  int<lower=0> N;
  int<lower=0> T;
  array[T] real x;
  array[N, T] real y;
  real xbar;
}
parameters {
  array[N] real alpha;
  array[N] real beta;
  
  real mu_alpha;
  real mu_beta; // beta.c in original bugs model
  
  real<lower=0> sigmasq_y;
  real<lower=0> sigmasq_alpha;
  real<lower=0> sigmasq_beta;
}
transformed parameters {
  real<lower=0> sigma_y; // sigma in original bugs model
  real<lower=0> sigma_alpha;
  real<lower=0> sigma_beta;
  
  sigma_y = sqrt(sigmasq_y);
  sigma_alpha = sqrt(sigmasq_alpha);
  sigma_beta = sqrt(sigmasq_beta);
}
model {
  mu_alpha ~ normal(0, 100);
  mu_beta ~ normal(0, 100);
  sigmasq_y ~ inv_gamma(0.001, 0.001);
  sigmasq_alpha ~ inv_gamma(0.001, 0.001);
  sigmasq_beta ~ inv_gamma(0.001, 0.001);
  alpha ~ normal(mu_alpha, sigma_alpha); // vectorized
  beta ~ normal(mu_beta, sigma_beta); // vectorized
  for (n in 1 : N) {
    for (t in 1 : T) {
      y[n, t] ~ normal(alpha[n] + beta[n] * (x[t] - xbar), sigma_y);
    }
  }
}
generated quantities {
  real alpha0;
  alpha0 = mu_alpha - xbar * mu_beta;
}



Error in stanc(file = file, model_code = model_code, model_name = model_name,  : 
  failed to parse Stan model 'rats' due to the above error.
